In [1]:
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import requests

In [3]:
#get html from wiki page and create soup object
source = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
soup = BeautifulSoup(source.text, 'lxml')

#using soup object, iterate the .wikitable to get the data from the HTML page and store it into a list
data = []
columns = []
table = soup.find(class_='wikitable')
for index, tr in enumerate(table.find_all('tr')):
    section = []
    for td in tr.find_all(['th','td']):
        section.append(td.text.rstrip())
    
    #First row of data is the header
    if (index == 0):
        columns = section
    else:
        data.append(section)

#convert list into Pandas DataFrame
canada_df = pd.DataFrame(data = data,columns = columns)
canada_df.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                            
1         M2A      Not assigned                            
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

### Here is a full list with all postal codes
<br>
**Then we will remove the Not assigned and N/A parts**

In [4]:
canada_df = canada_df[canada_df['Borough'] != 'Not assigned']
canada_df.head(10)

Postal code           Borough                                  Neighborhood
2          M3A        North York                                     Parkwoods
3          M4A        North York                              Victoria Village
4          M5A  Downtown Toronto                    Regent Park / Harbourfront
5          M6A        North York             Lawrence Manor / Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government
8          M9A         Etobicoke                              Islington Avenue
9          M1B       Scarborough                               Malvern / Rouge
11         M3B        North York                                     Don Mills
12         M4B         East York              Parkview Hill / Woodbine Gardens
13         M5B  Downtown Toronto                      Garden District, Ryerson

instructions: More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

In [5]:
#check
canada_df.loc[canada_df['Postal code']=='M5A']

Postal code           Borough                Neighborhood
4         M5A  Downtown Toronto  Regent Park / Harbourfront

In [6]:
canada_df.loc[canada_df['Postal code']=='M9V']

Postal code    Borough                                       Neighborhood
143         M9V  Etobicoke  South Steeles / Silverstone / Humbergate / Jam...

**So here we gonna replace the / sign with , to match the desired table**

In [15]:
canada_df['Neighborhood']=canada_df['Neighborhood'].str.replace('/',',')
canada_df.head(10)

Postal code           Borough                                  Neighborhood
2          M3A        North York                                     Parkwoods
3          M4A        North York                              Victoria Village
4          M5A  Downtown Toronto                    Regent Park , Harbourfront
5          M6A        North York             Lawrence Manor , Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government
8          M9A         Etobicoke                              Islington Avenue
9          M1B       Scarborough                               Malvern , Rouge
11         M3B        North York                                     Don Mills
12         M4B         East York              Parkview Hill , Woodbine Gardens
13         M5B  Downtown Toronto                      Garden District, Ryerson

In [16]:
canada_df.reset_index(drop=True)

Postal code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
..          ...               ...   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                           Regent Park , Harbourfront  
3                    Lawrence Manor , Lawrence Heights  
4         Queen's Park , Ontario Provincial Government  
5                                     Islington Avenue  
6                                      Malvern , Rouge  
7                                            Don Mills  
8                     Parkview Hill , Woodbine Gardens  
9                             Garden District, Ryerson  
10                                           Glencairn  
11   West Deane Park , Princess Gardens , Martin Gr...  
12            Rouge Hill , Port Union , Highland Creek  
13                                           Don Mills  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Eringate , Bloordale Gardens , Old Burnhamthor...  
18                 Guildwood , Morningside , West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  Central Bay Street  
25                                            Christie  
26                                           Cedarbrae  
27                                   Hillcrest Village  
28   Bathurst Manor , Wilson Heights , Downsview North  

In [18]:
np.shape(canada_df)

(103, 3)

In [20]:
from geopy.geocoders import Nominatim

import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

---
### Part 2

In [21]:
!wget -q -O "toronto_coordinates.csv" http://cocl.us/Geospatial_data
print('Coordinates downloaded!')
coors = pd.read_csv('toronto_coordinates.csv')

Coordinates downloaded!


In [22]:
coors.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [24]:
canada_df_temp = canada_df.set_index('Postal code')
coors_temp = coors.set_index('Postal Code')
toronto_df_coors = pd.concat([canada_df_temp, coors_temp], axis=1, join='inner')

toronto_df_coors.index.name = 'PostalCode'
toronto_df_coors.reset_index(inplace=True)

print(toronto_df_coors.shape)
toronto_df_coors.head(10)

(103, 5)


PostalCode           Borough                                  Neighborhood  \
0        M3A        North York                                     Parkwoods   
1        M4A        North York                              Victoria Village   
2        M5A  Downtown Toronto                    Regent Park , Harbourfront   
3        M6A        North York             Lawrence Manor , Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government   
5        M9A         Etobicoke                              Islington Avenue   
6        M1B       Scarborough                               Malvern , Rouge   
7        M3B        North York                                     Don Mills   
8        M4B         East York              Parkview Hill , Woodbine Gardens   
9        M5B  Downtown Toronto                      Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

---
### Part 3

In [25]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tl-toronto-neigh")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


**Use the technique previous in python visualization**

In [26]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, long, post, borough, neigh in zip(toronto_df_coors['Latitude'], toronto_df_coors['Longitude'], toronto_df_coors['PostalCode'], toronto_df_coors['Borough'], toronto_df_coors['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

**Here is all the datapoints, we should reduce the number adjusted our data**

In [28]:
toronto_boroughs = ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']
toronto_central_df = toronto_df_coors[toronto_df_coors['Borough'].isin(toronto_boroughs)].reset_index(drop=True)
#print(toronto_central_df.shape)
toronto_central_df.head()

PostalCode           Borough                                  Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park , Harbourfront   
1        M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government   
2        M5B  Downtown Toronto                      Garden District, Ryerson   
3        M5C  Downtown Toronto                                St. James Town   
4        M4E      East Toronto                                   The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [29]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, long, post, borough, neigh in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

---
## Final Project Part

### Description <br>
**Here we will try to find out the best area in Toronto to live for young people, old people, and the best area for business.<br> This is very useful for some real estate agencies to develop their business and consulting companies to help their clients.**

In [30]:
# First we will use the Foursquare API
CLIENT_ID = '5LCOWRSOOBERWFQWQOF11VSTPQQU12NQK5R4ETGVES1C3ELA' # your Foursquare ID
CLIENT_SECRET = 'B0OC10NHX3IYERVXXKK3UXSN0PPODFWRT1GU2DRD4P2UULTK' # your Foursquare Secret
VERSION = '20190330' # Foursquare API version
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [31]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['PostalCode', 'Borough', 'Neighborhood', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.shape)
venues_df.head(10)

(1624, 9)


PostalCode           Borough                Neighborhood  BoroughLatitude  \
0        M5A  Downtown Toronto  Regent Park , Harbourfront         43.65426   
1        M5A  Downtown Toronto  Regent Park , Harbourfront         43.65426   
2        M5A  Downtown Toronto  Regent Park , Harbourfront         43.65426   
3        M5A  Downtown Toronto  Regent Park , Harbourfront         43.65426   
4        M5A  Downtown Toronto  Regent Park , Harbourfront         43.65426   
5        M5A  Downtown Toronto  Regent Park , Harbourfront         43.65426   
6        M5A  Downtown Toronto  Regent Park , Harbourfront         43.65426   
7        M5A  Downtown Toronto  Regent Park , Harbourfront         43.65426   
8        M5A  Downtown Toronto  Regent Park , Harbourfront         43.65426   
9        M5A  Downtown Toronto  Regent Park , Harbourfront         43.65426   

   BoroughLongitude                         VenueName  VenueLatitude  \
0        -79.360636                  Roselle Desserts      43.653447   
1        -79.360636                     Tandem Coffee      43.653559   
2        -79.360636                Morning Glory Cafe      43.653947   
3        -79.360636            Cooper Koo Family YMCA      43.653249   
4        -79.360636               Body Blitz Spa East      43.654735   
5        -79.360636                    Impact Kitchen      43.656369   
6        -79.360636                   Corktown Common      43.655618   
7        -79.360636                The Extension Room      43.653313   
8        -79.360636  The Distillery Historic District      43.650244   
9        -79.360636            Figs Breakfast & Lunch      43.655675   

   VenueLongitude         VenueCategory  
0      -79.362017                Bakery  
1      -79.361809           Coffee Shop  
2      -79.361149        Breakfast Spot  
3      -79.358008   Distribution Center  
4      -79.359874                   Spa  
5      -79.356980            Restaurant  
6      -79.356211                  Park  
7      -79.359725  Gym / Fitness Center  
8      -79.359323         Historic Site  
9      -79.364503        Breakfast Spot

In [32]:
# Now we can check the type of venues with its counts here
venues_df.groupby(['PostalCode', 'Borough', 'Neighborhood'])['VenueName'].count()

PostalCode  Borough           Neighborhood                                                                                                    
M4E         East Toronto      The Beaches                                                                                                           5
M4K         East Toronto      The Danforth West , Riverdale                                                                                        43
M4L         East Toronto      India Bazaar , The Beaches West                                                                                      20
M4M         East Toronto      Studio District                                                                                                      40
M4N         Central Toronto   Lawrence Park                                                                                                         3
M4P         Central Toronto   Davisville North                                                             

**The above shows us, with a zip code, in each area, the number of venues**

In [33]:
print("there are ",len(venues_df['VenueCategory'].unique())," types of venues in Toronto")

there are  232  types of venues in Toronto


### Now we will create the matrix for data mining

In [34]:
# introduce dummy variables
toronto_central_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_central_onehot['PostalCode'] = venues_df['PostalCode'] 
toronto_central_onehot['Borough'] = venues_df['Borough'] 
toronto_central_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_central_onehot.columns[-3:]) + list(toronto_central_onehot.columns[:-3])
toronto_central_onehot = toronto_central_onehot[fixed_columns]

print(toronto_central_onehot.shape)
toronto_central_onehot.head(10)

(1624, 235)


PostalCode           Borough               Neighborhoods  Airport  \
0        M5A  Downtown Toronto  Regent Park , Harbourfront        0   
1        M5A  Downtown Toronto  Regent Park , Harbourfront        0   
2        M5A  Downtown Toronto  Regent Park , Harbourfront        0   
3        M5A  Downtown Toronto  Regent Park , Harbourfront        0   
4        M5A  Downtown Toronto  Regent Park , Harbourfront        0   
5        M5A  Downtown Toronto  Regent Park , Harbourfront        0   
6        M5A  Downtown Toronto  Regent Park , Harbourfront        0   
7        M5A  Downtown Toronto  Regent Park , Harbourfront        0   
8        M5A  Downtown Toronto  Regent Park , Harbourfront        0   
9        M5A  Downtown Toronto  Regent Park , Harbourfront        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   
5                   0             0               0                0   
6                   0             0               0                0   
7                   0             0               0                0   
8                   0             0               0                0   
9                   0             0               0                0   

   Airport Terminal  American Restaurant     ...       Theme Restaurant  \
0                 0                    0     ...                      0   
1                 0                    0     ...                      0   
2                 0                    0     ...                      0   
3                 0                    0     ...                      0   
4                 0                    0     ...                      0   
5                 0                    0     ...                      0   
6                 0                    0     ...                      0   
7                 0                    0     ...                      0   
8                 0                    0     ...                      0   
9                 0                    0     ...                      0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      0              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   
5                 0      0              0                              0   
6                 0      0              0                              0   
7                 0      0              0                              0   
8                 0      0              0                              0   
9                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Women's Store  \
0                 0                      0         0              0   
1                 0                      0         0              0   
2                 0                      0         0              0   
3                 0                      0         0              0   
4                 0                      0         0              0   
5                 0                      0         0              0   
6                 0                      0         0              0   
7                 0                      0         0              0   
8                 0                      0         0              0   
9                 0                      0         0  

**Here the 235 cols are 232 types of venues plus post code, borough and neighborhood**

In [35]:
# to avoid sparse matrix, we will use the mean freq instead of occurance
toronto_central_venues_freq = toronto_central_onehot.groupby(['PostalCode', 'Borough', 'Neighborhoods']).mean().reset_index()
print(toronto_central_venues_freq.shape)
toronto_central_venues_freq.head()

(39, 235)


PostalCode          Borough                    Neighborhoods  Airport  \
0        M4E     East Toronto                      The Beaches      0.0   
1        M4K     East Toronto    The Danforth West , Riverdale      0.0   
2        M4L     East Toronto  India Bazaar , The Beaches West      0.0   
3        M4M     East Toronto                  Studio District      0.0   
4        M4N  Central Toronto                    Lawrence Park      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant     ...       Theme Restaurant  \
0               0.0             0.000000     ...                    0.0   
1               0.0             0.023256     ...                    0.0   
2               0.0             0.000000     ...                    0.0   
3               0.0             0.050000     ...                    0.0   
4               0.0             0.000000     ...                    0.0   

   Toy / Game Store     Trail  Train Station  Vegetarian / Vegan Restaurant  \
0               0.0  0.200000            0.0                            0.0   
1               0.0  0.023256            0.0                            0.0   
2               0.0  0.000000            0.0                            0.0   
3               0.0  0.000000            0.0                            0.0   
4               0.0  0.000000            0.0                            0.0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Women's Store  \
0               0.0                    0.0     0.000            0.0   
1               0.0                    0.0     0.000            0.0   
2               0.0                    0.0     0.000            0.0   
3               0.0                    0.0     0.025            0.0   
4               0.0                    0.0     0.000            0.0   

   Yoga Studio  
0     0.000000  
1     0.023256  
2     0.000000  
3     0.025000  
4     0.000000  

[5 rows x 235 columns]

In [44]:
num_top_venues = 15

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['PostalCode', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = toronto_central_venues_freq['PostalCode']
neighborhoods_venues_sorted['Borough'] = toronto_central_venues_freq['Borough']
neighborhoods_venues_sorted['Neighborhoods'] = toronto_central_venues_freq['Neighborhoods']

for ind in np.arange(toronto_central_venues_freq.shape[0]):
    row_categories = toronto_central_venues_freq.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
neighborhoods_venues_sorted

PostalCode           Borough  \
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
25        M5S  Downtown Toronto   
33        M6K      West Toronto   
15        M5C  Downtown Toronto   
3         M4M      East Toronto   
26        M5T  Downtown Toronto   
14        M5B  Downtown Toronto   
6         M4R   Central Toronto   
19        M5J  Downtown Toronto   
29        M5X  Downtown Toronto   
18        M5H  Downtown Toronto   
21        M5L  Downtown Toronto   
16        M5E  Downtown Toronto   
20        M5K  Downtown Toronto   
17        M5G  Downtown Toronto   
28        M5W  Downtown Toronto   
36        M6S      West Toronto   
13        M5A  Downtown Toronto   
37        M7A  Downtown Toronto   
7         M4S   Central Toronto   
2         M4L      East Toronto   
35        M6R      West Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
22        M5N   Central Toronto   
34        M6P      West Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
23        M5P   Central Toronto   
8         M4T   Central Toronto   
10        M4W  Downtown Toronto   
31        M6H      West Toronto   
0         M4E      East Toronto   
9         M4V   Central Toronto   
11        M4X  Downtown Toronto   
24        M5R   Central Toronto   
12        M4Y  Downtown Toronto   
38        M7Y      East Toronto   

                                        Neighborhoods  \
27  CN Tower , King and Spadina , Railway Lands , ...   
32                          Little Portugal , Trinity   
25                    University of Toronto , Harbord   
33     Brockton , Parkdale Village , Exhibition Place   
15                                     St. James Town   
3                                     Studio District   
26        Kensington Market , Chinatown , Grange Park   
14                           Garden District, Ryerson   
6                                  North Toronto West   
19  Harbourfront East , Union Station , Toronto Is...   
29            First Canadian Place , Underground city   
18                         Richmond , Adelaide , King   
21                    Commerce Court , Victoria Hotel   
16                                        Berczy Park   
20          Toronto Dominion Centre , Design Exchange   
17                                 Central Bay Street   
28                                     Stn A PO Boxes   
36                                Runnymede , Swansea   
13                         Regent Park , Harbourfront   
37       Queen's Park , Ontario Provincial Government   
7                                          Davisville   
2                     India Bazaar , The Beaches West   
35                            Parkdale , Roncesvalles   
1                       The Danforth West , Riverdale   
30                                           Christie   
22                                           Roselawn   
34                     High Park , The Junction South   
4                                       Lawrence Park   
5                                    Davisville North   
23                           Forest Hill North & West   
8                        Moore Park , Summerhill East   
10                                           Rosedale   
31                      Dufferin , Dovercourt Village   
0                                         The Beaches   
9   Summerhill West , Rathnelly , South Hill , For...   
11                       St. James Town , Cabbagetown   
24              The Annex , North Midtown , Yorkville   
12                               Church and Wellesley   
38              Business reply mail Processing CentrE   

      1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
27          Airport Service        Airport Lounge      Airport Terminal   
32                      Bar            Restaurant           Men's Store   
25                     Café                Bakery                   Bar   
33                   

In [47]:
kclusters = 7

toronto_central_venues_freq_clustering = toronto_central_venues_freq.drop(['PostalCode', 'Borough', 'Neighborhoods'], 1)

kmeans = KMeans(n_clusters=kclusters, random_state=100).fit(toronto_central_venues_freq_clustering)

toronto_central_clustered_df = toronto_central_df
toronto_central_clustered_df['Cluster'] = kmeans.labels_

toronto_central_clustered_df = toronto_central_clustered_df.join(neighborhoods_venues_sorted.drop(['Borough', 'Neighborhoods'], 1).set_index('PostalCode'), on='PostalCode')
toronto_central_clustered_df.sort_values(['Cluster'] + freqColumns, inplace=True)
toronto_central_clustered_df

PostalCode           Borough  \
32        M5V  Downtown Toronto   
11        M6J      West Toronto   
27        M5S  Downtown Toronto   
14        M6K      West Toronto   
3         M5C  Downtown Toronto   
17        M4M      East Toronto   
30        M5T  Downtown Toronto   
2         M5B  Downtown Toronto   
36        M5X  Downtown Toronto   
16        M5L  Downtown Toronto   
13        M5K  Downtown Toronto   
6         M5G  Downtown Toronto   
34        M5W  Downtown Toronto   
28        M6S      West Toronto   
1         M7A  Downtown Toronto   
26        M4S   Central Toronto   
15        M4L      East Toronto   
25        M6R      West Toronto   
12        M4K      East Toronto   
7         M6G  Downtown Toronto   
19        M5N   Central Toronto   
18        M4N   Central Toronto   
20        M4P   Central Toronto   
21        M5P   Central Toronto   
29        M4T   Central Toronto   
33        M4W  Downtown Toronto   
31        M4V   Central Toronto   
35        M4X  Downtown Toronto   
24        M5R   Central Toronto   
37        M4Y  Downtown Toronto   
38        M7Y      East Toronto   
4         M4E      East Toronto   
10        M5J  Downtown Toronto   
8         M5H  Downtown Toronto   
22        M6P      West Toronto   
23        M4R   Central Toronto   
5         M5E  Downtown Toronto   
0         M5A  Downtown Toronto   
9         M6H      West Toronto   

                                         Neighborhood   Latitude  Longitude  \
32  CN Tower , King and Spadina , Railway Lands , ...  43.628947 -79.394420   
11                          Little Portugal , Trinity  43.647927 -79.419750   
27                    University of Toronto , Harbord  43.662696 -79.400049   
14     Brockton , Parkdale Village , Exhibition Place  43.636847 -79.428191   
3                                      St. James Town  43.651494 -79.375418   
17                                    Studio District  43.659526 -79.340923   
30        Kensington Market , Chinatown , Grange Park  43.653206 -79.400049   
2                            Garden District, Ryerson  43.657162 -79.378937   
36            First Canadian Place , Underground city  43.648429 -79.382280   
16                    Commerce Court , Victoria Hotel  43.648198 -79.379817   
13          Toronto Dominion Centre , Design Exchange  43.647177 -79.381576   
6                                  Central Bay Street  43.657952 -79.387383   
34                                     Stn A PO Boxes  43.646435 -79.374846   
28                                Runnymede , Swansea  43.651571 -79.484450   
1        Queen's Park , Ontario Provincial Government  43.662301 -79.389494   
26                                         Davisville  43.704324 -79.388790   
15                    India Bazaar , The Beaches West  43.668999 -79.315572   
25                            Parkdale , Roncesvalles  43.648960 -79.456325   
12                      The Danforth West , Riverdale  43.679557 -79.352188   
7                                            Christie  43.669542 -79.422564   
19                                           Roselawn  43.711695 -79.416936   
18                                      Lawrence Park  43.728020 -79.388790   
20                                   Davisville North  43.712751 -79.390197   
21                           Forest Hill North & West  43.696948 -79.411307   
29                       Moore Park , Summerhill East  43.689574 -79.383160   
33                                           Rosedale  43.679563 -79.377529   
31  Summerhill West , Rathnelly , South Hill , For...  43.686412 -79.400049   
35                       St. James Town , Cabbagetown  43.667967 -79.367675   
24              The Annex , North Midtown , Yorkville  43.672710 -79.405678   
37                               Church and Wellesley  43.665860 -79.383160   
38              Business reply mail Processing CentrE  43.662744 -79.321558   
4                                         The Beaches  43.676357 -79.29303

In [48]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster in zip(toronto_central_clustered_df['Latitude'], toronto_central_clustered_df['Longitude'], toronto_central_clustered_df['PostalCode'], toronto_central_clustered_df['Borough'], toronto_central_clustered_df['Neighborhood'], toronto_central_clustered_df['Cluster']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [55]:
toronto_central_clustered_df[toronto_central_clustered_df['Cluster']==1]

PostalCode       Borough Neighborhood   Latitude  Longitude  Cluster  \
4        M4E  East Toronto  The Beaches  43.676357 -79.293031        1   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
4                   Pub           Coffee Shop          Neighborhood   

  4th Most Common Venue          ...           6th Most Common Venue  \
4                 Trail          ...             Dumpling Restaurant   

  7th Most Common Venue        8th Most Common Venue 9th Most Common Venue  \
4            Donut Shop  Eastern European Restaurant         Deli / Bodega   

  10th Most Common Venue 11th Most Common Venue 12th Most Common Venue  \
4       Doner Restaurant                Dog Run      Electronics Store   

  13th Most Common Venue 14th Most Common Venue 15th Most Common Venue  
4    Distribution Center         Discount Store                  Diner  

[1 rows x 21 columns]

### Conclusion:
<br>
#### Here I create 7 clusters based on the 15 most important features of each area.<br>
- For the red points, they are the area that are similar which have **no unique** features from other area.<br>
- We can find for the orange points, there are more art facilities(Music venue, theater, performing arts venue...) and people seems doing more exercise here (more Athletics & Sports, parks...). So the orange area should be the place for people who want to live a healthy life with parks and art venues around, for example, people retired.<br>
- For cluster 3 which is the light blue point, the top 1-14 are all restaurant related venues. In other words, this area is a really diversified foodies area. For people who want to open a fancy restaurant, this is the place he should go to.<br>
- For cluster 2 which are the darker blue points, this is a tourists place with lots of hotels, restaurant , gyms. So for business owners who are targeting on tourists, this is definetly the best area to go.
- For cluster 5 which is the yellow point. This area has many bars, pubs, Italian restaurants, etc. So this should be an area that young people love to hang out and date. Young people who concerns night life important or people who are trying to find a best dating place should look into this area.
- For cluster 1 which is the purple, this is a typical living area for people not that rich since the top venues including the trail and discount store.